src/execution_plan/optimizations/traverse_order.c

```diff
@@ -5,74 +5,229 @@
 */
 
 #include "./traverse_order.h"
+#include "../../util/arr.h"
 #include "../../util/vector.h"
+#include "../../util/rmalloc.h"
+#include <assert.h>
 
-/* Given a set of algebraic expressions and the entire filter tree,
- * suggest traversal entry point to be either the first expression or the
- * last expression, in the future we'll want to be able to begin traversal from 
- * any expression. */
-TRAVERSE_ORDER determineTraverseOrder(const FT_FilterNode *filterTree,
-                                      AlgebraicExpression **exps,
-                                      size_t expCount) {
+#define T 1     // Transpose penalty.
+#define F 4 * T // Filter score.
 
-    if(expCount == 1 || !filterTree) {
-        return TRAVERSE_ORDER_FIRST;
+typedef AlgebraicExpression** Arrangement;
+
+// Create a new arrangement.
+Arrangement Arrangement_New(uint size) {
+    return rm_malloc(sizeof(AlgebraicExpression*) * size);
+}
+
+// Clone arrangement.
+Arrangement Arrangement_Clone(const Arrangement arrangement, uint size) {
+    assert(arrangement);
+    Arrangement clone = Arrangement_New(size);
+    memcpy(clone, arrangement, sizeof(AlgebraicExpression*) * size);
+    return clone;
+}
+
+// Print arrangement.
+void Arrangement_Print(Arrangement arrangement, uint size) {
```

Comment:
I don't mind keeping debug routines around, but am commenting here in case you intended to remove this!

Reply:
Thanks, for the time-being let's keep it